### ISSUE TITLE
Background indexing is too slow / demanding

### ISSUE THREAD

**Issue by MahdiBM** [swiftlang/sourcekit-lsp#1837]
### Swift version

Swift 6.0.2 release

### Platform

macOS 15

### Editor

VS Code

### Description

Background indexing looks to be 2-5x slower than Xcode's mechanism, and that's already a low bar to meet.

Perhaps we can have some benchmarks for sourcekit-lsp (though a package like [ordo-one/package-benchmark](https://github.com/ordo-one/package-benchmark)), then we can both make sure that this doesn't regress any further, and also try to optimize it slowly but surely.

We'll need to discuss how to set up the benchmarking, if that's desirable.

### Steps to Reproduce

Enable background indexing on a big-enough project.

**Comment by MahdiBM** [comment]
I wonder if some `Task.yield()`s are going to help or not.

**Comment by ahoppen** [comment]
No, there aren’t any few lines that may speed up background indexing here, it’s more architectural than that. A couple things that are getting addressed:
- When we prepare a target for indexing, we currently launch a `swift build --prepare-target-for-indexing` process, which needs to resolve the package manifest. Resolving the package manifest may take 1 or 2 seconds (especially when plugins are involved), which is a significant chunk if the actual target preparation only takes 1 second. This will be addressed by having a long-lived BSP server for SwiftPM projects that keeps the build graph in memory and can use it for target preparation. https://github.com/swiftlang/swift-build/pull/782 is laying the groundwork for that
- We used to launch a new compiler process to index each file. This resulted in duplicated work for loading modules for every file in a target. https://github.com/swiftlang/sourcekit-lsp/pull/2293 supports indexing multiple files in a single invocation, which should yield benefits here
- There may be more work that can be done inside the compiler or the build system to skip unnecessary work during indexing, but most of the low hanging fruit should already be picked there.
